### DISCUSSION TITLE
How to do package versions based on TFM condition for CentralPackageVersionManagement?

### DISCUSSION THREAD

**Discussion by japj** [NuGet/Home#11773] [Q&A]
How would one use conditions on the TargetFramework with CentralPackageVersionManagement (using Directory.Packages.props).
In case you have projects with different TargetFramework and needing to use a different version based on the TargetFramework.

Example Directory.Packages.Props
```xml
<Project>
  <ItemGroup>
    <PackageVersion Condition=" '$(TargetFramework)' == 'netstandard2.0' " Include="MSTest.TestAdapter" Version="1.1.0" />
    <PackageVersion Condition=" '$(TargetFramework)' == 'net6.0' " Include="MSTest.TestAdapter" Version="3.1.0" />
  </ItemGroup>
</Project>
```

As far as I can tell the `Directory.Packages.Props` file is imported before the TargetFramework Property is set in a normal csproj file causing the conditions to evaluate to empty and thus not working.

**Comment by erdembayar** [comment]
Fyi @jeffkl

**Comment by jeffkl** [comment]
In MSBuild, all properties and imports are evaluated first.  Items are then evaluated second.  Since `PackageVersion` is an item, it is evaluated after all properties even if it is declared before a property.  The following works as expected for me:

## ClassLibrary1.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>net472;net6.0</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" />
  </ItemGroup>
</Project>
```

## Directory.Packages.props
```xml
<Project>
  <PropertyGroup>
    <!-- Enable central package management -->
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
  </PropertyGroup>
  <ItemGroup>
    <PackageVersion Include="Newtonsoft.Json" Version="12.0.1" Condition="'$(TargetFramework)' == 'net472'" />
    <PackageVersion Include="Newtonsoft.Json" Version="13.0.1" Condition="'$(TargetFramework)' != 'net472'" />
  </ItemGroup>
</Project>
```

![image](https://user-images.githubusercontent.com/17556515/166257202-1511cbf7-a084-4497-80cd-6ad7f3a403a4.png)

If its not working as expected, please do a command-line based restore and attach [`msbuild.binlog`](https://www.msbuildlog.com/):

```
msbuild /Target:Restore /BinaryLogger
```

**Reply by nvmkpk** [reply]
Not true, you are multi targeting, in this case it is working because the TFM is set before loading the project for specific TFM. If you specify single TFM using `TargetFramework` element, it will fail.
